File: Assets/CloudCode/ValidateAndSetUsername.js
// ValidateAndSetUsername.js
// Deploy to UGS Cloud Code: Dashboard → Cloud Code → Scripts → Create
//
// This server-side script validates a candidate arena username for uniqueness
// and, if valid, atomically claims it for the calling player.
//
// Username index is stored in Cloud Save Game Data (Private Custom Items):
//   Entity: "username_index", Key: normalised username, Value: playerId
//
// Parameters:
//   username (string) — the candidate display name
//
// Returns:
//   { success: true }                         — username claimed
//   { success: false, error: "reason" }       — validation or uniqueness failure

const { DataApi } = require("@unity-services/cloud-save-1.4");

const MIN_LENGTH = 2;
const MAX_LENGTH = 16;
const CUSTOM_ID = "username_index";

module.exports = async ({ params, context, logger }) => {
  const { projectId, playerId } = context;
  const username = params.username;

  // ── Client-side validation (belt-and-suspenders) ──

  if (!username || typeof username !== "string") {
    return { success: false, error: "Username is required." };
  }

  const trimmed = username.trim();

  if (trimmed.length < MIN_LENGTH) {
    return {
      success: false,
      error: `Name must be at least ${MIN_LENGTH} characters.`,
    };
  }

  if (trimmed.length > MAX_LENGTH) {
    return {
      success: false,
      error: `Name must be at most ${MAX_LENGTH} characters.`,
    };
  }

  if (!isUsernameSafe(trimmed)) {
    return { success: false, error: "Name contains invalid characters." };
  }

  // ── Single service-scoped client (handles both player and game data) ──
  const cloudSave = new DataApi(context);

  // ── Check if this player already has a username ──

  let existingData;
  try {
    existingData = await cloudSave.getItems(projectId, playerId, ["arena_v1"]);
  } catch (e) {
    logger.error("Failed to read player data: " + e.message);
    return { success: false, error: "Server error reading player data." };
  }

  if (existingData?.data?.results) {
    for (const item of existingData.data.results) {
      if (item.key === "arena_v1" && item.value) {
        let arenaData;
        try {
          arenaData =
            typeof item.value === "string"
              ? JSON.parse(item.value)
              : item.value;
        } catch (_) {
          /* ignore parse error */
        }

        if (arenaData && arenaData.usernameCreated && arenaData.arenaUsername) {
          return { success: false, error: "You already have a username." };
        }
      }
    }
  }

  // ── Check uniqueness via Game Data (private custom items) ──

  const nameKey = trimmed.toLowerCase();
  let nameTaken = false;
  let oldNameKey = null; // Track if this player already owns a different name.

  try {
    const indexResult = await cloudSave.getPrivateCustomItems(
      projectId,
      CUSTOM_ID,
    );
    logger.info(
      "Username index read OK. Results: " +
        JSON.stringify(indexResult?.data?.results?.length ?? 0),
    );

    if (indexResult?.data?.results) {
      for (const item of indexResult.data.results) {
        // Check if this player owns a different name (from a previous reset).
        if (item.value === playerId && item.key !== nameKey) {
          oldNameKey = item.key;
        }
        if (item.key === nameKey) {
          if (item.value && item.value !== playerId) {
            nameTaken = true;
          }
          // If value === playerId, this player already owns it (crash retry).
        }
      }
    }
  } catch (e) {
    // 404 means the entity doesn't exist yet — name is available.
    const status = e?.response?.status || e?.status;
    if (status !== 404) {
      logger.error(
        "Failed to check username index: status=" +
          status +
          " message=" +
          e.message,
      );
      return {
        success: false,
        error: "Server error checking name availability.",
      };
    }
    logger.info("Username index entity not found (404) — first username ever.");
  }

  if (nameTaken) {
    return { success: false, error: "That name is already taken." };
  }

  // ── Release old name if this player is switching ──

  if (oldNameKey) {
    try {
      await cloudSave.deletePrivateCustomItem(projectId, CUSTOM_ID, oldNameKey);
      logger.info(`Released old username index entry: "${oldNameKey}"`);
    } catch (e) {
      // Non-fatal — log but continue.
      logger.error("Failed to release old username: " + e.message);
    }
  }

  // ── Claim the name ──

  try {
    await cloudSave.setPrivateCustomItem(projectId, CUSTOM_ID, {
      key: nameKey,
      value: playerId,
    });
    logger.info(
      `Wrote username index: entity="${CUSTOM_ID}", key="${nameKey}", value="${playerId}"`,
    );
  } catch (e) {
    logger.error("Failed to write username index: " + e.message);
    return { success: false, error: "Server error claiming name." };
  }

  // ── Write to the player's arena data ──

  try {
    let arenaData = {};
    if (existingData?.data?.results) {
      for (const item of existingData.data.results) {
        if (item.key === "arena_v1" && item.value) {
          try {
            arenaData =
              typeof item.value === "string"
                ? JSON.parse(item.value)
                : item.value;
          } catch (_) {
            arenaData = {};
          }
        }
      }
    }

    arenaData.arenaUsername = trimmed;
    arenaData.usernameCreated = true;

    await cloudSave.setItem(projectId, playerId, {
      key: "arena_v1",
      value: JSON.stringify(arenaData),
    });
  } catch (e) {
    logger.error("Failed to write player arena data: " + e.message);
    return {
      success: false,
      error: "Name claimed but profile update failed. Reopen the Arena.",
    };
  }

  logger.info(`Player ${playerId} claimed username "${trimmed}"`);
  return { success: true };
};

function isUsernameSafe(name) {
  if (!name || name.trim() !== name) return false;
  for (let i = 0; i < name.length; i++) {
    const c = name.charCodeAt(i);
    if (c < 32) return false; // control characters
    if (c === 60 || c === 62) return false; // < >
  }
  return true;
}

File: ArenaModule/ArenaModule/Handlers/ValidateAndSetUsername.cs
// ArenaModule — ValidateAndSetUsername handler
// Port of ValidateAndSetUsername.js to C# Cloud Code Module.
//
// Validates a candidate arena username for uniqueness and atomically claims it.
// Username index stored in Cloud Save Custom Data entity "username_index".

using System.Text.Json;
using Microsoft.Extensions.Logging;
using Unity.Services.CloudCode.Apis;
using Unity.Services.CloudCode.Core;
using Unity.Services.CloudCode.Shared;
using Unity.Services.CloudSave.Model;

namespace ArenaModule.Handlers;

public class ValidateAndSetUsername
{
    private const int MinLength = 2;
    private const int MaxLength = 16;
    private const string CustomEntityId = "username_index";
    private const string ArenaDataKey = "arena_v1";

    private readonly ILogger<ValidateAndSetUsername> _logger;

    public ValidateAndSetUsername(ILogger<ValidateAndSetUsername> logger)
    {
        _logger = logger;
    }

    [CloudCodeFunction("ValidateAndSetUsername")]
    public async Task<UsernameResult> Execute(
        IExecutionContext ctx,
        IGameApiClient api,
        string username)
    {
        // ── Client-side validation (belt-and-suspenders) ──

        if (string.IsNullOrWhiteSpace(username))
            return Fail("Username is required.");

        var trimmed = username.Trim();

        if (trimmed.Length < MinLength)
            return Fail($"Name must be at least {MinLength} characters.");

        if (trimmed.Length > MaxLength)
            return Fail($"Name must be at most {MaxLength} characters.");

        if (!IsUsernameSafe(trimmed))
            return Fail("Name contains invalid characters.");

        // ── Parallel read: fetch both player data AND username index concurrently ── // FIXED: parallelize I/O to cut latency
        var nameKey = trimmed.ToLowerInvariant();
        ArenaSaveDataServer? arenaData = null;
        bool nameTaken = false;
        string? oldNameKey = null;

        try
        {
            // FIXED: fetch both independent datasets concurrently instead of sequentially
            var playerTask = api.CloudSaveData.GetItemsAsync(
                ctx, ctx.AccessToken, ctx.ProjectId, ctx.PlayerId,
                new List<string> { ArenaDataKey });

            var indexTask = api.CloudSaveData.GetCustomItemsAsync(
                ctx, ctx.ServiceToken, ctx.ProjectId, CustomEntityId);

            await Task.WhenAll(playerTask, indexTask);

            var playerItems = playerTask.Result;
            var indexResult = indexTask.Result;

            // ── Process player data ──
            if (playerItems?.Data?.Results != null)
            {
                foreach (var item in playerItems.Data.Results)
                {
                    if (item.Key == ArenaDataKey && item.Value != null)
                    {
                        var raw = item.Value.ToString();
                        if (!string.IsNullOrEmpty(raw))
                            arenaData = JsonSerializer.Deserialize<ArenaSaveDataServer>(raw);
                    }
                }
            }

            if (arenaData is { UsernameCreated: true, ArenaUsername: not null })
                return Fail("You already have a username.");

            // ── Process username index ──
            if (indexResult?.Data?.Results != null)
            {
                foreach (var item in indexResult.Data.Results)
                {
                    var value = item.Value?.ToString() ?? "";

                    // This player owns a different name (previous reset)
                    if (value == ctx.PlayerId && item.Key != nameKey)
                        oldNameKey = item.Key;

                    if (item.Key == nameKey)
                    {
                        if (!string.IsNullOrEmpty(value) && value != ctx.PlayerId)
                            nameTaken = true;
                    }
                }
            }
        }
        catch (ApiException ex) when (ex.Response?.StatusCode == System.Net.HttpStatusCode.NotFound)
        {
            _logger.LogInformation("Username index entity not found (404) — first username ever.");
        }
        catch (Exception ex)
        {
            _logger.LogError("Failed during parallel read: {Error}", ex.Message);
            return Fail("Server error during validation.");
        }

        if (nameTaken)
            return Fail("That name is already taken.");

        // ── Release old name if switching ──

        if (oldNameKey != null)
        {
            try
            {
                await api.CloudSaveData.DeleteCustomItemAsync(
                    ctx, ctx.ServiceToken, ctx.ProjectId, CustomEntityId, oldNameKey);
                _logger.LogInformation("Released old username index entry: \"{OldKey}\"", oldNameKey);
            }
            catch (Exception ex)
            {
                _logger.LogError("Failed to release old username: {Error}", ex.Message);
                // Non-fatal — continue
            }
        }

        // ── Claim the name ──

        try
        {
            await api.CloudSaveData.SetCustomItemAsync(
                ctx, ctx.ServiceToken, ctx.ProjectId, CustomEntityId,
                new SetItemBody(nameKey, ctx.PlayerId));
            _logger.LogInformation("Wrote username index: entity=\"{Entity}\", key=\"{Key}\", value=\"{PlayerId}\"",
                CustomEntityId, nameKey, ctx.PlayerId);
        }
        catch (Exception ex)
        {
            _logger.LogError("Failed to write username index: {Error}", ex.Message);
            return Fail("Server error claiming name.");
        }

        // ── Write to the player's arena data ──

        try
        {
            arenaData ??= new ArenaSaveDataServer();
            arenaData.ArenaUsername = trimmed;
            arenaData.UsernameCreated = true;

            var json = JsonSerializer.Serialize(arenaData);
            await api.CloudSaveData.SetItemAsync(
                ctx, ctx.AccessToken, ctx.ProjectId, ctx.PlayerId,
                new SetItemBody(ArenaDataKey, json));
        }
        catch (Exception ex)
        {
            _logger.LogError("Failed to write player arena data: {Error}", ex.Message);
            return Fail("Name claimed but profile update failed. Reopen the Arena.");
        }

        _logger.LogInformation("Player {PlayerId} claimed username \"{Name}\"", ctx.PlayerId, trimmed);
        return new UsernameResult { Success = true };
    }

    // ── Helpers ──

    private static UsernameResult Fail(string error)
        => new() { Success = false, Error = error };

    private static bool IsUsernameSafe(string name)
    {
        if (string.IsNullOrEmpty(name) || name.Trim() != name) return false;
        foreach (char c in name)
        {
            if (c < 32) return false;          // control characters
            if (c == '<' || c == '>') return false; // angle brackets
        }
        return true;
    }
}
